### DISCUSSION TITLE
About non-idempotent capsules

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#243] [General]
I need to settle my ReArch knowledge on this regard. I think I already undrestood this but I need confirmation before adopting an specific approach on my current project:
- Does non-idempotent capsules would participate non-stop on the rebuild process once they are readed (or used) once, even if they stop being used at a given point? I guess yes.
- Will be always like this?

Thanks

**Comment by GregoryConrad** [comment]
As soon as a non-idempotent capsule is read for the first time, it remains in the container until the end of the container’s life. This is because it relies on the state or something from outside of the containers control, so it’s not safe to dispose (unlike idempotent capsules).

> Will be always like this?

likely, yes, that was a fundamental design decision 

if you need state that gets disposed:
- have it return an explicit reset() function, OR
- maybe you should be using ephemeral state instead

**Reply by busslina** [reply]
Thanks, it's fine for me to be like this, in fact is what I need. I have one main state(\*) and several sub-states(\*) and I wanted to "boot" the app logic by reading the main state capsule which will handle all the logic (main state processing and sub-states initialization).

This leads to some doubts about idempotent capsules that I will ask on another thread.

Thanks


(*) State machines
